### PR TITLE
Add SpecialRecord Type and Updated Config

### LINKED ISSUE #598: Implement Validator Rotation and Minimal Slashing
@djrtwo introduced [validator rotation and minimal slashing](https://github.com/ethereum/eth2.0-specs/pull/17) for 2.1 today. This issue will address those functionalities within our `casper` packages. It includes:
- Handle special messages validators can sign outside of normal consensus
- Implement routine to add validator into the pool
- Track special messages in `SpecialObjects`
- Process these special messages during crystallized state transition
- Implement validator shuffling in and out of active set, also perform withdraws after withdrawal period has ended

I'll work on this after v0.0.0 completes

### PR DESCRIPTION
**Pull request by terenc3t**
This partially fixes #598

What got changed in this PR:
- Added new proto type `SpecialRecord` This is used for validator rotation and slashing
- Updated config and improved comments with the latest spec

### CODE REVIEW

File: beacon-chain/params/config.go [hunk 1]
```diff
@@ -10,53 +10,59 @@ var env = "default"
...
-	CycleLength                 uint64    // CycleLength is the beacon chain cycle length in slots.
-	SlotDuration                uint64    // SlotDuration in seconds.
+	CycleLength                 uint64    // CycleLength is one beacon chain cycle length in slots.
+	SlotDuration                uint64    // SlotDuration is how many seconds are in a single slot.
 	MinCommiteeSize             uint64    // MinDynastyLength is the slots needed before dynasty transition happens.
 	DefaultEndDynasty           uint64    // DefaultEndDynasty is the upper bound of dynasty. We use it to track queued and exited validators.
 	MinDynastyLength            uint64    // MinCommiteeSize is the minimal number of validator needs to be in a committee.
 	BaseRewardQuotient          uint64    // BaseRewardQuotient is used to calculate validator per-slot interest rate.
-	SqrtDropTime                uint64    // SqrtDropTime is a constant to reflect time it takes to cut offline validators’ deposits by 39.4%.
+	SqrtEDropTime               uint64    // SqrtEDropTime is a constant to reflect time it takes to cut offline validators’ deposits by 39.4%.
```
Comment (prestonvanloon) [inline beacon-chain/params/config.go]: What does it mean though? What does the E stand for?

Comment (terenc3t) [reply beacon-chain/params/config.go]: E is this guys 
https://en.wikipedia.org/wiki/E_(mathematical_constant)

Comment (rauljordan) [reply beacon-chain/params/config.go]: Rename to Exp perhaps as that is more standard in computing

File: beacon-chain/params/config.go [hunk 2]
```diff
@@ -10,53 +10,59 @@ var env = "default"
...
+	DepositSize:               new(big.Int).Div(big.NewInt(32), big.NewInt(int64(1e18))),
+	SlotDuration:              uint64(8),
+	MinCommiteeSize:           uint64(3),
+	DefaultEndDynasty:         uint64(999999999999999999),
+	MinDynastyLength:          uint64(256),
+	BaseRewardQuotient:        uint64(32768),
+	SqrtEDropTime:             uint64(65536),
+	WithdrawalPeriod:          uint64(128),
+	MaxValidatorChurnQuotient: uint64(32),
+	LogOutMessage:             "LOGOUT",
```
Comment (rauljordan) [inline beacon-chain/params/config.go]: Shall we use enums instead?

File: beacon-chain/utils/shuffle.go [hunk 3]
```diff
@@ -12,8 +12,11 @@ import (
 // ShuffleIndices returns a list of pseudorandomly sampled
 // indices. This is used to use to select attesters and proposers.
 func ShuffleIndices(seed common.Hash, validatorList []uint32) ([]uint32, error) {
-	if len(validatorList) > params.GetConfig().MaxValidators {
-		return nil, errors.New("Validator count has exceeded MaxValidator Count")
+	// since we are consuming 3 bytes of entropy at a time in the loop,
```
Comment (rauljordan) [inline beacon-chain/utils/shuffle.go]: Capitalize Since

**Comment by prestonvanloon** [comment]
I added a few commits to address the following pattern:
```
    for i := 0; i < params.GetConfig().ModuloBias+1; i++ {
        validators = append(validators, &pb.ValidatorRecord{StartDynasty: 0, EndDynasty: params.GetConfig().DefaultEndDynasty})
    }
```
this was creating 16M copies of the same pb when a single copy would suffice. The original memory footprint for some of these tests was nearly 3gb.

```
Showing nodes accounting for 3097.24MB, 99.72% of 3105.92MB total                                                                                                       
Dropped 27 nodes (cum <= 15.53MB)                                                                                                                                       
      flat  flat%   sum%        cum   cum%                                                                                                                              
 3032.23MB 97.63% 97.63%  3096.24MB 99.69%  github.com/prysmaticlabs/prysm/beacon-chain/types.TestNewDynastyRecalculationsInvalid                                       
   65.01MB  2.09% 99.72%    65.01MB  2.09%  github.com/prysmaticlabs/prysm/beacon-chain/casper.ActiveValidatorIndices                                                   
         0     0% 99.72%    64.01MB  2.06%  github.com/prysmaticlabs/prysm/beacon-chain/casper.ShuffleValidatorsToCommittees                                            
         0     0% 99.72%    64.01MB  2.06%  github.com/prysmaticlabs/prysm/beacon-chain/types.(*CrystallizedState).newDynastyRecalculations
         0     0% 99.72%  3102.75MB 99.90%  testing.tRunner
```

After creating a slice with the backing array of the length we expected (save some allocs on resizing the array) and using the same pointer for every validator in the list (it was the same data anyway), we could reduce the footprint to <200MB.

```
Showing nodes accounting for 197.69MB, 99.73% of 198.22MB total
Dropped 2 nodes (cum <= 0.99MB)
Showing top 10 nodes out of 22
      flat  flat%   sum%        cum   cum%
  128.01MB 64.58% 64.58%   192.02MB 96.87%  github.com/prysmaticlabs/prysm/beacon-chain/types.TestNewDynastyRecalculationsInvalid
   64.51MB 32.54% 97.12%    64.51MB 32.54%  github.com/prysmaticlabs/prysm/beacon-chain/casper.ActiveValidatorIndices
       2MB  1.01% 98.13%        2MB  1.01%  github.com/prysmaticlabs/prysm/beacon-chain/types.initialValidators
    1.16MB  0.58% 98.71%     1.16MB  0.58%  runtime/pprof.StartCPUProfile
```
This is kind of an example of the impact of having all of the validators in memory. It would likely use about 3Gb of ram for the 16M validators. The active subset would be much smaller, but we should keep this in mind for constrained environments (raspberry pi, old android phones, etc).

**Comment by djrtwo** [comment]
Note that max validators is closer to 4M (assuming 32 eth deposits and ~130M eth supply cap. 

The max size that the shuffling algorithm can handle is that 16M due to the entropy being consumed 3 bytes at a time (rand max)  

So the max ram requirements of storing validators in memory would be closer to 0.75GB rather than 4

**Pull request merged**